File: src/leetcode/java/p916/Solution.java
package leetcode.java.p916;
// https://leetcode.com/problems/word-subsets/submissions/

import java.util.LinkedList;
import java.util.List;

public class Solution {
    public List<String> wordSubsets(String[] words1, String[] words2) {
        int[] filterVec = new int[26];
        for (String word : words2) {
            int[] vec = convertToFrequencyVec(word);
            for (int i = 0; i < 26; ++i) {
                filterVec[i] = Math.max(filterVec[i], vec[i]);
            }
        }

        List<String> ans = new LinkedList<>();
        for (String word : words1) {
            if (isUniversal(word, filterVec))
                ans.add(word);
        }
        return ans;
    }

    private boolean isUniversal(final String word, final int[] filter) {
        int[] freqVec = convertToFrequencyVec(word);
        for (int i = 0; i < 26; ++i) {
            if (freqVec[i] < filter[i])
                return false;
        }
        return true;
    }

    private int[] convertToFrequencyVec(final String word) {
        int[] vec = new int[26];
        for (final char c : word.toCharArray()) {
            ++vec[c - 'a'];
        }
        return vec;
    }
}

File: src/leetcode/cpp/p916/solution.cpp
// https://leetcode.com/problems/word-subsets/
/***
 * String
 * Time: O(n|word|+m|word|)
 * Space: O(n)
 */
#include <bits/stdc++.h>
using namespace std;

#define FAST_IO ios_base::sync_with_stdio(0), cin.tie(0);
#define bitcount __builtin_popcount
#define countZeros __builtin_ctz
using ll= long long;
using ld= long double;
using pii= pair<int,int>;
using pll= pair<ll,ll>;
using vi= vector<int>;
using vll= vector<ll>;
using vs= vector<string>;
using vpll= vector<pll>;
using vvi= vector<vi>;
using vvll= vector<vll>;
using vvpll= vector<vpll>;
using maxHeap= priority_queue<int, vector<int>, less<int>>;
using minHeap= priority_queue<int, vector<int>, greater<int>>;

class Solution {
public:
    vs wordSubsets(vs& words1, vs& words2) {
        vvi words1Vec;
        for (string word : words1) {
            words1Vec.push_back(toVector(word));
        }
        vi maxVec(26);
        for (string word : words2) {
            vi vec= toVector(word);
            for (int i=0; i < 26; ++i) {
                maxVec[i]= max(maxVec[i], vec[i]);
            }
        }

        vs ans;
        for (int i=0; i < words1.size(); ++i) {
            if (isSubset(words1Vec[i], maxVec)) {
                ans.push_back(words1[i]);
            }
        }
        return ans;
    }

private:
    vi toVector(string word) {
        vi ret(26);
        for (char c : word) {
            ret[c-'a']++;
        }
        return ret;
    }

    bool isSubset(vi& a, vi& b) {
        for (int i=0; i < a.size(); ++i) {
            if (a[i] < b[i]) return false;
        }
        return true;
    }
};